File: day18/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	current := Pos{x: 0, y: 0}
	current2 := Pos{x: 0, y: 0}
	area := 0
	perimeter := 0
	area2 := 0
	perimeter2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.Split(scanner.Text(), " ")
		if len(line) != 3 {
			log.Fatalln("Invalid line")
		}
		{
			dir := line[0]
			num, err := strconv.Atoi(line[1])
			if err != nil {
				log.Fatalln(err)
			}
			current = move(current, dir, num)
			switch dir {
			case "L":
				area -= current.y * num
			case "R":
				area += current.y * num
			}
			perimeter += num
		}
		{
			line2 := strings.Trim(line[2], "(#)")
			if len(line2) != 6 {
				log.Fatalln("Invalid line2")
			}
			dir := line2[5]
			num64, err := strconv.ParseInt(string(line2[:5]), 16, 64)
			if err != nil {
				log.Fatalln(err)
			}
			num := int(num64)
			current2 = move2(current2, dir, num)
			switch dir {
			case '2':
				area2 -= current2.y * num
			case '0':
				area2 += current2.y * num
			}
			perimeter2 += num
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	if perimeter%2 != 0 {
		log.Fatalln("Perimeter not aligned to grid")
	}
	area = abs(area)
	halfPerimeter := perimeter / 2
	fmt.Println("Part 1:", area+halfPerimeter+1)

	if perimeter2%2 != 0 {
		log.Fatalln("Perimeter not aligned to grid")
	}
	area2 = abs(area2)
	halfPerimeter2 := perimeter2 / 2
	fmt.Println("Part 2:", area2+halfPerimeter2+1)
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func move2(p Pos, dir byte, num int) Pos {
	switch dir {
	case '3':
		p.y -= num
	case '1':
		p.y += num
	case '2':
		p.x -= num
	case '0':
		p.x += num
	default:
		log.Fatalln("Invalid dir")
	}
	return p
}

func move(p Pos, dir string, num int) Pos {
	switch dir {
	case "U":
		p.y -= num
	case "D":
		p.y += num
	case "L":
		p.x -= num
	case "R":
		p.x += num
	default:
		log.Fatalln("Invalid dir")
	}
	return p
}

type (
	Pos struct {
		x, y int
	}
)

func getBounds(p, tl, br Pos) (Pos, Pos) {
	return Pos{
			x: min(tl.x, p.x),
			y: min(tl.y, p.y),
		}, Pos{
			x: max(br.x, p.x),
			y: max(br.y, p.y),
		}
}

File: day18/src/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut current = Pos { x: 0, y: 0 };
    let mut current2 = Pos { x: 0, y: 0 };
    let mut area = 0;
    let mut area2 = 0;
    let mut perimeter = 0;
    let mut perimeter2 = 0;

    for line in reader.lines() {
        let line = line?;
        let (p1, p2) = line.rsplit_once(' ').ok_or("Invalid line")?;
        {
            let (dir_str, num_str) = p1.split_once(' ').ok_or("Invalid line")?;
            let dir = dir_str.as_bytes()[0];
            let num = num_str.parse::<isize>()?;
            current = move_dir(current, dir, num);
            match dir {
                b'L' => area -= current.y * num,
                b'R' => area += current.y * num,
                _ => {}
            }
            perimeter += num;
        }
        {
            let p2 = &p2[2..8];
            let dir = p2.as_bytes()[5] - b'0';
            let num = isize::from_str_radix(&p2[..5], 16)?;
            current2 = move_dir2(current2, dir, num);
            match dir {
                2 => area2 -= current2.y * num,
                0 => area2 += current2.y * num,
                _ => {}
            }
            perimeter2 += num;
        }
    }

    if perimeter % 2 != 0 {
        return Err("Perimeter not aligned to grid".into());
    }
    if perimeter2 % 2 != 0 {
        return Err("Perimeter not aligned to grid".into());
    }
    area = area.abs();
    area2 = area2.abs();

    println!(
        "Part 1: {}\nPart 2: {}",
        area + perimeter / 2 + 1,
        area2 + perimeter2 / 2 + 1,
    );

    Ok(())
}

struct Pos {
    x: isize,
    y: isize,
}

fn move_dir(mut p: Pos, dir: u8, num: isize) -> Pos {
    match dir {
        b'U' => p.y -= num,
        b'D' => p.y += num,
        b'L' => p.x -= num,
        b'R' => p.x += num,
        _ => {}
    }
    p
}

fn move_dir2(mut p: Pos, dir: u8, num: isize) -> Pos {
    match dir {
        3 => p.y -= num,
        1 => p.y += num,
        2 => p.x -= num,
        0 => p.x += num,
        _ => {}
    }
    p
}
